vpr/src/analytical_place/ap_netlist.h

```diff
@@ -0,0 +1,188 @@
+
+    /**
+     * @brief Create or return an existing pin in the netlist
+     *
+     *  @param port_id  The port this pin is associated with
+     *  @param port_bit The bit index of the pin in the port
+     *  @param net_id   The net the pin drives/sinks
+     *  @param pin_type The type of the pin (driver/sink)
+     *  @param is_const Indicates whether the pin holds a constant value (e.g.
+     *                  vcc/gnd)
+     */
+    APPinId create_pin(const APPortId port_id, BitIndex port_bit, const APNetId net_id, const PinType pin_type, bool is_const = false);
+
+    /**
+     * @brief Create an empty, or return an existing net in the netlist
+     *
+     *  @param name The unique name of the net
+     */
+    APNetId create_net(const std::string& name);
+
+private: // Private Members
+    /*
+     * Netlist compression / optimization
+     */
+
+    /// @brief Removes invalid components and reorders them
+    void clean_blocks_impl(const vtr::vector_map<APBlockId, APBlockId>& block_id_map) override;
+    void clean_ports_impl(const vtr::vector_map<APPortId, APPortId>& port_id_map) override;
+    void clean_pins_impl(const vtr::vector_map<APPinId, APPinId>& pin_id_map) override;
+    void clean_nets_impl(const vtr::vector_map<APNetId, APNetId>& net_id_map) override;
+
+    void rebuild_block_refs_impl(const vtr::vector_map<APPinId, APPinId>& pin_id_map, const vtr::vector_map<APPortId, APPortId>& port_id_map) override;
+    void rebuild_port_refs_impl(const vtr::vector_map<APBlockId, APBlockId>& block_id_map, const vtr::vector_map<APPinId, APPinId>& pin_id_map) override;
+    void rebuild_pin_refs_impl(const vtr::vector_map<APPortId, APPortId>& port_id_map, const vtr::vector_map<APNetId, APNetId>& net_id_map) override;
+    void rebuild_net_refs_impl(const vtr::vector_map<APPinId, APPinId>& pin_id_map) override;
+
+    /// @brief Shrinks internal data structures to required size to reduce
+    ///        memory consumption
+    void shrink_to_fit_impl() override;
+
+    /*
+     * Component removal
+     */
+    void remove_block_impl(const APBlockId blk_id) override;
+    void remove_port_impl(const APPortId port_id) override;
+    void remove_pin_impl(const APPinId pin_id) override;
+    void remove_net_impl(const APNetId net_id) override;
+
+    /*
+     * Sanity checks
+     */
+    // Verify the internal data structure sizes match
+    bool validate_block_sizes_impl(size_t num_blocks) const override;
+    bool validate_port_sizes_impl(size_t num_ports) const override;
+    bool validate_pin_sizes_impl(size_t num_pins) const override;
+    bool validate_net_sizes_impl(size_t num_nets) const override;
+
+private: // Private Data
+    /// @brief Molecule of each block
+    vtr::vector_map<APBlockId, const t_pack_molecule*> block_molecules_;
+    /// @brief Type of each block
+    vtr::vector_map<APBlockId, APBlockType> block_types_;
```

Comment:
Could infer the mobility of each block from block_locs_ with a member function.

Reply:
Could also make an enum that is part of the APFixedBlockLoc to represent mobility status.

Reply:
Looking through my future PRs on AP; the mobility is queried much more often than the fixed block locations. Having that as a quick (cache local) lookup is more valuable to me in my opinion.